Add ContinuousStates and -Observables

EwoutH opened the pull request:
### Summary
Adds `ContinuousObservable` to mesa_signals, enabling agent states that change continuously over time with automatic threshold detection and signal emission. Demonstrates the feature with an enhanced Wolf-Sheep predator-prey model where energy depletes continuously rather than at discrete time steps.

### Motive
Agent-based models often need to represent continuously changing states like energy depletion, resource growth, or temperature changes. Currently, modelers must manually calculate state changes at each step or schedule discrete update events, leading to:
- Balancing trade-offs between update frequency and performance
- Difficulty detecting exact threshold crossings
- No standardized patterns for time-varying states
- Complex coordination of multiple changing states

This addresses discussion #2529 (Continuous States) and provides a foundational building block toward the behavioral framework outlined in discussion #2538, enabling more realistic agent behaviors driven by internal state dynamics.

### Implementation
**Core Components:**

1. **`ContinuousObservable` descriptor**: Extends `Observable` to track values that change over time according to a rate function. Uses lazy evaluation - values are only recalculated when accessed, making it efficient even with many agents.
2. **`ContinuousState` helper class**: Internal state tracker storing the current value, last update time, rate function, and threshold set. Handles time-based calculation and threshold crossing detection.
3. **Threshold management**: `HasObservables.add_threshold()` provides a clean API for registering callbacks when values cross specific thresholds (upward or downward). Uses the existing signal/observer pattern - thresholds emit `"threshold_crossed"` signals.
4. **Time integration**: Uses the new unified `model.time` (see #2903).

**Key design decisions:**
- Lazy evaluation prevents unnecessary calculations when states aren't accessed
- Threshold storage uses a set (values only), callbacks managed through signal subscriptions to avoid duplicate invocations
- Linear integration for simplicity; extensible to more complex methods
- Fully integrated with existing `Computed` properties and dependency tracking

### API usage examples
```python
# Continuous value with rate per time unit
energy = ContinuousObservable(initial_value: float, rate_func: (value, elapsed, agent) -> float)

# Set/adjust value (emits "change" + checks thresholds)
self.energy = 120.0
self.energy += 5.0

# Thresholds (fires "threshold_crossed" with signal.threshold, signal.direction)
self.add_threshold("energy", 0.0, callback)

# Subscribe to signals
self.observe("energy", "change", on_change)
self.observe("energy", "threshold_crossed", on_threshold)

# Computed properties that depend on observables
is_hungry = Computable()
self.is_hungry = Computed(lambda: self.energy < 50.0)
```

### Usage Examples
**Basic continuous energy depletion:**
```python
class Wolf(Agent, HasObservables):
    energy = ContinuousObservable(
        initial_value=100.0,
        rate_func=lambda value, elapsed, agent: -agent.metabolic_rate
    )
    
    def __init__(self, model):
        super().__init__(model)
        self.metabolic_rate = 0.5
        self.energy = 100.0
        
        # Die when energy reaches zero
        self.add_threshold("energy", 0.0, self._on_death)
    
    def _on_death(self, signal):
        if signal.direction == "down":
            self.remove()
```

**Reactive behaviors with computed properties:**
```python
class Animal(Agent, HasObservables):
    energy = ContinuousObservable(
        initial_value=100.0,
        rate_func=lambda value, elapsed, agent: -agent.metabolic_rate
    )
    is_hungry = Computable()
    
    def __init__(self, model):
        super().__init__(model)
        self.energy = 100.0
        # Computed property automatically updates when energy changes
        self.is_hungry = Computed(lambda: self.energy < 50.0)
    
    def move(self):
        if self.is_hungry:
            # Hunt actively when hungry
            self.seek_food()
        else:
            # Conserve energy when not hungry
            self.wander()
```

**Enhanced Wolf-Sheep model**: The updated example demonstrates continuous energy dynamics, threshold-triggered death/starvation modes, and behavior switching based on computed hunger states. Energy depletes continuously between steps rather than in discrete chunks, creating more realistic predator-prey dynamics.

### Additional Notes
- **Tests**: Comprehensive test suite (19 tests) covering basic functionality, signals, edge cases (numpy floats, zero elapsed time, exact thresholds), and integration scenarios
- **Backward compatible**: Existing mesa_signals code unchanged; `ContinuousObservable` is an additive feature
- **Performance**: Lazy evaluation ensures minimal overhead; states only recalculate when accessed
- **Future extensions**: Foundation for more complex behavioral patterns (needs-based architectures, homeostatic agents) discussed in #2538
- **Documentation**: Wolf-Sheep example serves as practical demonstration; could expand docs with more use cases

This provides the first of two foundational building blocks identified for Mesa's behavioral framework: reactive state management with continuous dynamics. The second block (action/task scheduling) can build naturally on this foundation.

EwoutH committed:
mesa_signals: add ContinuousObservable for time-varying agent states

Add ContinuousObservable class that extends mesa_signals to support agent properties that change continuously over time. This enables modeling realistic phenomena like energy depletion, resource growth, and gradual state changes without manual step-by-step updates.

Key features:
- Lazy evaluation: values only recalculated when accessed, improving performance
- Threshold detection: automatic callbacks when values cross specified thresholds in either direction
- Configurable rate functions: support linear, exponential, or custom change patterns
- Full integration with existing mesa_signals reactive system: works seamlessly with Observable, Computable, and dependency tracking
- Time-aware: integrates with Mesa's time system for accurate elapsed time calculations

This provides a foundational building block for modeling needs-based architectures, continuous resource systems, and time-dependent agent behaviors while maintaining the composability and reactivity of mesa_signals.

fix imports

EwoutH committed:
Enhance wolf-sheep with continuous energy and reactive behaviors

Update the wolf-sheep predation example to demonstrate mesa_signals and ContinuousObservable capabilities. Energy now depletes continuously over time rather than in discrete steps, and agent behaviors react automatically to energy state changes.

Changes:
- Convert Animal base class to use ContinuousObservable for energy tracking with configurable metabolic rates
- Add computed properties (is_hungry, can_reproduce) that automatically update based on energy levels
- Implement threshold-triggered events for death (energy reaches zero) and survival mode (critical hunger)
- Enhance decision logic: sheep prioritize grass when hungry, wolves hunt actively when hungry
- Add activity-based metabolic rates (movement costs more energy than resting)
- Improve GrassPatch to use Observable and react to state changes with signal handlers
- Expand data collection to track average energy levels across populations

The example now demonstrates key reactive programming patterns: observable state management, computed properties for decision making, threshold-based event triggering, and automatic dependency tracking. This makes the model more realistic (continuous dynamics) while reducing boilerplate code (declarative state management).

fix agent imports

EwoutH committed:
Add time-source fallback mechanism for ContinuousObservable

Implement flexible time retrieval in ContinuousObservable to work with different Mesa time management approaches. The _get_time() method tries multiple sources in priority order:

1. model.simulator.time (DEVS/continuous models)
2. model.time (if explicitly set)
3. model.steps (fallback for discrete models)

This workaround enables ContinuousObservable to function correctly regardless of whether a model uses discrete event simulation, custom time tracking, or standard step-based progression.

Obviously this should be fixed structurally. See #2228

EwoutH committed:
implement __set__ for ContinuousObservable to properly initialize state

The ContinuousObservable descriptor was inheriting Observable's __set__ method, which stored raw numeric values instead of ContinuousState objects. This caused AttributeError when __get__ tried to access state.last_update on a float/numpy.float64.

The new __set__ method creates a ContinuousState wrapper on first assignment and updates the existing state on subsequent assignments. This ensures the private attribute always contains a properly structured state object with value, last_update, rate_func, and thresholds attributes.

Fixes initialization of continuous observables in agent __init__ methods where energy and other time-varying properties are set.

EwoutH committed:
Fix threshold subscription to prevent duplicate callbacks

Changes the ContinuousObservable threshold system from storing callbacks
directly to using a signal-based subscription pattern:

- Change `_thresholds` from dict to set (stores threshold values only)
- Update `add_threshold()` to check for existing subscriptions before
  subscribing to prevent duplicate callback invocations
- Separate concerns: thresholds define WHICH values to watch, observers
  define WHO to notify when crossed
- Add ValueError when attempting to add threshold to non-ContinuousObservable

This fixes an issue where registering multiple thresholds with the same
callback would cause it to be called multiple times per threshold crossing.

EwoutH committed:
Add tests for ContinuousObservable

Adds test_continuous_observables.py with 19 tests covering the full  functionality of continuous state management in mesa_signals:

Basic functionality:
- Constant and variable rate depletion over time
- Manual value setting (e.g., energy boost from eating)
- Time-based lazy evaluation

Signal emissions:
- Change notifications when values update
- Threshold crossing detection (upward and downward)
- Multiple threshold management
- No signals when values don't change

Edge cases:
- Zero elapsed time (no spurious updates)
- Multiple accesses at same time point (single calculation)
- Exact threshold values (inclusive boundary handling)
- Negative values (energy going below zero)
- NumPy float compatibility (fixes AttributeError with random values)

Integration:
- Computed properties based on continuous observables
- Batch agent creation with numpy arrays
- Both DEVSimulator (float time) and ABMSimulator (integer steps)
- Wolf-sheep predator-prey scenario with death thresholds

All tests use proper discrete event scheduling through Mesa's simulator rather than manual time manipulation, ensuring realistic simulation behavior.

EwoutH commented:
I'm a bit mentally exhausted, was a big push to get this finished. I will write up some review points and context later this weekend.

The individual commit messages hopefully contain useful details.

For now:

1. Maybe the new `ContinuousObservable` and `ContinuousState` classes should go into a separate file instead of `mesa_signals.py`.
2. Yes, I know using ContiniousStates with an ABMSimulator (in the wolf_sheep example) doesn't make that much sense in the current implementation. Future work.
3. We really should re-address https://github.com/projectmesa/mesa/discussions/2228 sooner than later.

Curious on initial thoughts!

quaquel commented:
I think this idea is very cool. I hope to find time next week to take a closer look.

One quick question: how is threshold management squared with the lazy evaluation of ContinousStates/ContinuousObservables?

EwoutH committed:
fix: Make ContinuousObservable thresholds instance-specific

BREAKING: Thresholds are now stored per-instance instead of per-class.

Previously, ContinuousObservable stored thresholds at the class level in a shared `_thresholds` set. This caused all instances to check the same threshold values, preventing agents from having individual thresholds based on their specific parameters (e.g., a wolf with 100 starting energy couldn't have a different critical threshold than one with 50 starting energy).

Changes:
- Move threshold storage from descriptor to ContinuousState instances
- Change threshold structure from set to dict: {threshold_value: set(callbacks)}
- Update add_threshold() to access and modify instance-level state
- Ensure threshold initialization happens before adding thresholds
- Add direction parameter to all threshold_crossed signal emissions

This allows each agent to maintain its own set of thresholds, enabling instance-specific reactive behaviors like `self.add_threshold("energy", self.starting_energy * 0.25, callback)`.

EwoutH committed:
test_continuous_observable: Add multi-agent test cases

quaquel commented on mesa/experimental/mesa_signals/mesa_signal.py:
```diff
@@ -472,6 +472,189 @@ def _mesa_notify(self, signal: AttributeDict):
         self.subscribers[observable][signal_type] = active_observers
 
+    def add_threshold(self, observable_name: str, threshold: float, callback: Callable):
+        """Convenience method for adding thresholds."""
```
just nitpicking, but this docstring can use some further elaboration beyond the one line summary. For example, it might be good to give a short explanation of what a treshold is and how it only applies to continuous observables.

quaquel commented on mesa/experimental/mesa_signals/mesa_signal.py:
```diff
@@ -472,6 +472,189 @@ def _mesa_notify(self, signal: AttributeDict):
+        Returns:
+            List of (threshold_value, direction) tuples for crossed thresholds
+        """
+        crossed = []
```
can you indicate whether you are only interested in moving above or moving below, or do you get any crossing of the threshold?

quaquel reviewed (commented):
I don't see anything obviously wrong here. However, it's been a while since I looked at this code, so I might take another look at this PR annd the rest of signals next week. In the meantime, three clarifying questions.

EwoutH commented:
Threshold detection for non-linear rate functions is absolutely not trivial. I explored some options here:
- https://github.com/projectmesa/mesa/discussions/2854

I'm not sure which direction I'm leaning personally yet.

tpike3 commented:
@coderabbitai full review

EwoutH commented:
I'm fine with reviewers using AI tools to help them review a PR, but I'm not triaging AI comments. So @tpike3 please triage the comments and resolve the ones that are not highly relevant in this stage in het PR review process.

EwoutH committed:
Refactor ContinuousObservable to use model.time directly

Replaces calls to the internal _get_time method with direct access to instance.model.time in ContinuousObservable. Removes the _get_time method, simplifying time retrieval and assuming model.time is always available.

quaquel committed:
Merge branch 'main' into ContiniousStates

quaquel commented on mesa/examples/advanced/wolf_sheep/agents.py:
```diff
@@ -35,33 +79,39 @@ def spawn_offspring(self):
         """Execute one step of the animal's behavior."""
-        # Move to random neighboring cell
+        # Move to neighboring cell (uses more energy than standing still)
+        self.metabolic_rate = 1.0  # Movement costs more energy
```
I don't understand this code. You first chance it to 1 and then back to 0.5, however, Move does not seem to touch the updating of energy so this seems to do nothing at all? 

If I am wrong, it highlights something else: the implicitness of how all this works and that might be a cause of concern as well.

quaquel commented:
Ok, then I am going to close this PR to keep the open PRs in check and focussed on stuff we actually want to move forward.

EwoutH commented:
I would like to keep this open as draft. I might rebase.

The thorough cleanup is appreciated though!

quaquel commented:
closing again. Can always be reopend or picked up again if needed.
